File: container-with-most-water/obzva.cpp
/**
 * 풀이
 * - container의 밑변이 넓고 높이가 높을 수록 저장하는 물의 양이 많습니다
 * - 따라서 밑변이 가장 넓은 container부터 탐색하면 탐색 효율을 높일 수 있다는 생각을 했습니다
 * - 양 끝에서부 two pointer를 이용하여 탐색을 시작합니다
 * - lo, hi 중에서 높이가 더 낮은 pointer를 안쪽으로 이동시킵니다
 * - 왜냐하면 높이가 더 낮은 pointer를 이동시켰을 때 기존 container보다 더 높이가 높은 container를 만들 수 있는 가능성이 생기기 때문입니다
 * 
 * Big O
 * - N: 주어진 배열 height의 크기
 * 
 * - Time complexity: O(N)
 *   - 배열 height를 조회하므로 전체 실행시간 또한 N에 비례하여 선형적으로 증가합니다
 * 
 * - Space complexity: O(1)
 */

class Solution {
public:
    int maxArea(vector<int>& height) {
        int n = height.size();
        int lo = 0;
        int hi = n - 1;

        int res = 0;
        while (lo < hi) {
            int w = hi - lo;
            int h = min(height[lo], height[hi]);
            
            res = max(res, w * h);

            if (height[lo] > height[hi]) {
                --hi;
            } else {
                ++lo;
            }
        }

        return res;
    }
};

File: validate-binary-search-tree/obzva.go
/*
풀이
- BST의 속성을 이해한 후, 해당 속성을 검사하는 dfs 함수를 이용하면 풀이할 수 있습니다
Big O
- N: root 트리의 노드 개수
- Time complexity: O(N)
  - 모든 노드에 대해 최대 1번의 탐색이 필요합니다
- Space complexity: O(logN) (O(N) at worst)
  - check 함수의 재귀호출 스택 깊이는 트리의 높이에 비례하여 증가하므로 일반적으로 O(logN)의 공간복잡도를 가진다고 볼 수 있습니다
    하지만 트리가 심하게 치우친 경우 O(N)까지 커질 수 있습니다
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 const (
	MIN = -(2_147_483_648 + 1)
	MAX = 2_147_483_647 + 1
)

func isValidBST(root *TreeNode) bool {
	return check(root.Left, MIN, root.Val) && check(root.Right, root.Val, MAX)
}

/*
helper dfs function
*/

func check(node *TreeNode, min int, max int) bool {
	// base case
	if node == nil {
		return true
	}
	// node.val should be in the boundary (min, max)
	if !(min < node.Val && node.Val < max) {
		return false
	}
	// check for children nodes
	return check(node.Left, min, node.Val) && check(node.Right, node.Val, max)
}
